File: haskell/src/Day01.hs
module Day01 (part1, part2) where

import Data.List (permutations)

pick :: Int -> [a] -> [[a]]
pick 0 _ = [[]]
pick _ [] = []
pick n (x:xs) = map (x:) (pick (n-1) xs) ++ pick n xs

arrange :: Int -> [a] -> [[a]]
arrange n = concatMap permutations . pick n

part1 :: [Int] -> Int
part1 xs =
  let
    powerSetOfLength2 = arrange 2 xs
    sumTo2020 = filter (\(x:y:_) -> x + y == 2020) powerSetOfLength2
    result = map (\(x:y:_) -> x * y) sumTo2020
  in head result

part2 :: [Int] -> Int
part2 xs = let
    powerSetOfLength3 = arrange 3 xs
    sumTo2020 = filter (\(x:y:z:_) -> x + y + z == 2020) powerSetOfLength3
    result = map (\(x:y:z:_) -> x * y * z) sumTo2020
  in head result

File: elixir/lib/advent_of_code/day_01.ex
defmodule AdventOfCode.Day01 do

  def arrange(0, _), do: [[]]
  def arrange(_, []), do: []
  def arrange(n, [x | xs]) do
    (for l <- arrange(n - 1, xs), do: [x | l]) ++ arrange(n, xs)
  end

  def part1(xs) do
    powerSetOfLength2 = arrange(2, xs)
    sumTo2020 = Enum.filter(powerSetOfLength2, fn [x,y] -> x + y == 2020 end)
    hd Enum.map(sumTo2020, fn [x,y] -> x * y end)
  end

  def part2(xs) do
    powerSetOfLength3 = arrange(3, xs)
    sumTo2020 = Enum.filter(powerSetOfLength3, fn [x,y,z] -> x + y + z == 2020 end)
    hd Enum.map(sumTo2020, fn [x,y,z] -> x * y * z end)
  end
end
